File: pa2/commands.cpp
//SQL Command Handler
//By: Benjamin Estela

//Header Files
#include "commands.h"
using namespace std;

//Class Function Implementation
//Default Constructor
Commands::Commands()
{
    m_command = -1;
    m_manipulation = NONE_DATA;
    m_line = "";
    m_file = "";
    m_prevsql = "";
}

//Destructor (not fully implemented)
Commands::~Commands(){}

//CREATE DATABASE
void Commands::CreateDatabase()
{
    string tempStr;
    stringstream ss(m_line);
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, ';'); //tempStr is now the variable
    int len = tempStr.length();
    char fileName[len + 1];
    strcpy(fileName, tempStr.c_str());
    if(mkdir(fileName, 0777) == -1)
    {
        cerr << "!Failed to create database " << fileName << " because it already exists." << endl;
    }
    else
    {
        cout << "Database " << fileName << " created." << endl;
    }
    
}

//DROP DATABASE
void Commands::DropDatabase()
{
    string tempStr;
    stringstream ss(m_line);
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, ';'); //tempStr is now the variable
    int len = tempStr.length();
    char fileName[len + 1];
    strcpy(fileName, tempStr.c_str());
    //rmdir removes folder
    if(rmdir(fileName) == -1)
    {
        cerr << "!Failed to delete " << fileName << " because it does not exist." << endl;
    }
    else
    {
        cout << "Database " << fileName << " deleted." << endl;
    }
}

//USE
void Commands::Use()
{
    string tempStr;
    stringstream ss(m_line);
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, ';'); //tempStr is now the variable
    int len = tempStr.length();
    char folderName[len + 1];
    strcpy(folderName, tempStr.c_str());
    //chdir is basically cd in bash
    if(chdir(folderName) == -1)
    {
        //error handling that chdir's one folder back
        char prevDir[len + 4];
        strcpy(prevDir, "../");
        strcat(prevDir, folderName);
        if(chdir(prevDir) == -1)
        {
            cerr << "!Failed to access " << folderName << "." << endl;
        }
        else
        {
            cout << "Using database " << folderName << "." << endl;
        }
    }
    else
    {
        cout << "Using database " << folderName << "." << endl;
    }
    return;
}

//CREATE TABLE
void Commands::CreateTable()
{
    string tempStr;
    stringstream ss(m_line);
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, ' '); //tempStr is now the variable
    if(access(tempStr.c_str(), F_OK) != -1)
    {
        cerr << "!Failed to create table " << tempStr.c_str() << " because it already exists." << endl;
    }
    else
    {
        ofstream file;
        file.open(tempStr.c_str());
        cout << "Table " << tempStr.c_str() << " created." << endl;
        //Splits up parameters.
        getline(ss, tempStr, '(');
        getline(ss, tempStr, ';');
        string strWrite;
        strWrite = tempStr.substr(0, tempStr.size() - 1);
        //replace ", " with " | "
        size_t startPos = 0;
        string from = ", ";
        string to = " | "; //may need to fix?
        while((startPos = strWrite.find(from, startPos)) != string::npos)
        {
            strWrite.replace(startPos, from.length(), to);
            startPos += to.length();
        }
        file << strWrite << endl;
        file.close();
        
    }
    
    return;
}

//DROP TABLE
void Commands::DropTable()
{
    string tempStr;
    stringstream ss(m_line);
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, ';'); //tempStr is now the variable
    //remove deletes a file
    if(remove(tempStr.c_str()) != 0)
    {
        cerr << "!Failed to delete table " << tempStr.c_str() << " because it does not exist." << endl;
    }
    else
    {
        cout << "Table " << tempStr.c_str() << " deleted." << endl;
    }

    return;
}

//ALTER TABLE
void Commands::AlterTable()
{
    string tempStr;
    stringstream ss(m_line);
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, ' '); //tempStr is now the variable
    int len = tempStr.length();
    char tableName[len + 1];
    strcpy(tableName, tempStr.c_str());

    string toupperStr = m_line;
    transform(toupperStr.begin(), toupperStr.end(), toupperStr.begin(), ::toupper);

    //Can add more ALTER COMMANDS in the future
    //ADD is basically concat to the end of the table
    if(toupperStr.find("ADD") != string::npos)
    {
        getline(ss, tempStr, ' ');
        getline(ss, tempStr, ';');
        ofstream file;
        file.open(tableName, ios_base::app);
        if(tempStr.size() >= 1)
        {
            file << " | " << tempStr; //TODO: better parsing for more vars
        }
        
        file.close();
        cout << "Table " << tableName << " modified." << endl;
    }
    else
    {
        cerr << "!Failed to alter table" << tableName << "due to invalid command." << endl;
    }
    
    return;
}

//SELECT * FROM (Query All)
void Commands::SelectFromTable()
{
    string tempStr;
    string tableName;
    stringstream ss(m_line);
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, ' ');
    getline(ss, tableName, ';');
    //This function basically prints out everything in the file
    //* is used for all and can with more if statements, can be used for different queries
    if(m_line.find(" * ") != string::npos)
    {
        ifstream file;
        file.open(tableName);
        if(file.is_open())
        {
            SetFile(tableName);
            while(getline(file, tempStr))
            {
                if(tempStr.size() > 2) //test more
                {
                    cout << tempStr << endl;
                }
            }
        }
        else
        {
            cerr << "!Failed to query table " << tableName << " because it does not exist." << endl;
        }
        file.close();
    }

    return;
}

//SELECT (Query Specific)
void Commands::SelectTable()
{
    string tempStr;
    stringstream ss(m_line);
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, '\n');
    if(!tempStr.empty())
    {
        tempStr.pop_back();
        if(tempStr.back() == ' ')
        {
            tempStr.pop_back();
        }
    }
    //cout << "Query:" << tempStr << "+" << endl;
    SetPrevSQL(tempStr);
    SetManipulation(SELECT_DATA);
    return;
}

//INSERT INTO
void Commands::InsertIntoTable()
{
    string tempStr;
    stringstream ss(m_line);
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, ' '); //tempStr is now the variable
    int len = tempStr.length();
    char tableName[len + 1];
    strcpy(tableName, tempStr.c_str());

    string toupperStr = m_line;
    transform(toupperStr.begin(), toupperStr.end(), toupperStr.begin(), ::toupper);

    if(toupperStr.find("VALUES") != string::npos)
    {
        getline(ss, tempStr, '(');
        getline(ss, tempStr, ')');
        //parse values
        //removes whitespace
        tempStr.erase(remove_if(tempStr.begin(), tempStr.end(), isSpace()), tempStr.end());
        string strWrite;
        strWrite = tempStr.substr(0, tempStr.size());
        //replace "," with " | "
        size_t startPos = 0;
        string from = ",";
        string to = " | "; //may need to fix?
        while((startPos = strWrite.find(from, startPos)) != string::npos)
        {
            strWrite.replace(startPos, from.length(), to);
            startPos += to.length();
        }
        //open file
        ofstream file;
        file.open(tableName, ios_base::app);
        if(tempStr.size() >= 1)
        {
            file << strWrite << endl;
        }
        
        file.close();

        //can only handle 1 record at a time
        cout << "1 new record inserted." << endl;
    }
    else
    {
        cerr << "!Failed to insert into table" << tableName << "due to invalid command." << endl;
    }
    return;
}

//UPDATE
void Commands::UpdateTable()
{
    string tempStr;
    stringstream ss(m_line);
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, '\n'); //???
    tempStr.erase(remove_if(tempStr.begin(), tempStr.end(), isSpace()), tempStr.end());
    ifstream file;
    file.open(tempStr);
    if(file.is_open())
    {
        SetFile(tempStr);
        SetManipulation(UPDATE_DATA);
        //cout << "Table to update: " << tempStr << endl;
    }
    else
    {
        cerr << "!Failed to find table " << tempStr << " because it does not exist." << endl;
    }
    file.close();
    return;
}

//SET
void Commands::SetTable()
{
    string tempStr;
    stringstream ss(m_line);
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, '\n');
    if(!tempStr.empty())
    {
        tempStr.pop_back();
        if(tempStr.back() == ' ')
        {
            tempStr.pop_back();
        }
    }
    SetPrevSQL(tempStr);
    //cout << "set: " << tempStr << "+" << endl;
    return;
}

//FROM
void Commands::FromTable()
{
    string tempStr;
    stringstream ss(m_line);
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, '\n'); //???
    tempStr.erase(remove_if(tempStr.begin(), tempStr.end(), isSpace()), tempStr.end());
    ifstream file;
    file.open(tempStr);
    if(file.is_open())
    {
        SetFile(tempStr);
        SetManipulation(UPDATE_DATA);
        //cout << "Table to query: " << tempStr << endl;
    }
    else
    {
        cerr << "!Failed to find table " << tempStr << " because it does not exist." << endl;
    }
    file.close();
    return;
}

//WHERE
void Commands::WhereTable()
{
    if(GetManipulation() == UPDATE_DATA)
    {
        cout << "UPDATE WHERE?" << endl;
    }
    else if(GetManipulation() == DELETE_DATA)
    {
        cout << "DELETE WHERE?" << endl;
    }
    else if(GetManipulation() == SELECT_DATA)
    {
        cout << "SELECT WHERE?" << endl;
    }
    else
    {
        cerr << "!Failed to manipulate data as improper function was read in." << endl;
    }

    //With data manipulation complete, the table name does not need to be stored anymore.
    SetManipulation(NONE_DATA);
    SetFile("");
    SetPrevSQL("");
    return;
}

//DELETE FROM
void Commands::DeleteFromTable()
{
    //cout << "DELETE" << endl;
    string tempStr;
    stringstream ss(m_line);
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, ' ');
    getline(ss, tempStr, '\n');
    tempStr.erase(remove_if(tempStr.begin(), tempStr.end(), isSpace()), tempStr.end());
    ifstream file;
    file.open(tempStr);
    if(file.is_open())
    {
        SetFile(tempStr);
        SetManipulation(DELETE_DATA);
        //cout << "Table to delete: " << tempStr << endl;
    }
    else
    {
        cerr << "!Failed to find table " << tempStr << " because it does not exist." << endl;
    }
    file.close();
    return;
}

//Set Methods
void Commands::SetCommand(int command)
{
    m_command = command;
}

void Commands::SetManipulation(int manipulation)
{
    m_manipulation = manipulation;
}

void Commands::SetLine(string line)
{
    m_line = line;
}

void Commands::SetFile(string file)
{
    m_file = file;
}

void Commands::SetPrevSQL(string prevsql)
{
    m_prevsql = prevsql;
}

//Get Methods
int Commands::GetCommand()
{
    return m_command;
}

int Commands::GetManipulation()
{
    return m_manipulation;
}

string Commands::GetLine()
{
    return m_line;
}

string Commands::GetFile()
{
    return m_file;
}

string Commands::GetPrevSQL()
{
    return m_prevsql;
}

File: pa2/commands.py
#File: Command Handler
#By: Benjamin Estela
#Class: CS 457

#Header Files
#Modules
import os

#Enumeration Definition
class cmdName(enumerate):
    NONE = 0
    UPDATE = 1
    DELETE = 2
    SELECT = 3

#Class Definition
class CommandHandler:
    DataManipulation = cmdName.NONE
    TableName = ""
    PrevSQL = ""

#Free Function Definitions

#CREATE DATABASE
def CreateDatabase(line):
    dbName = line.split(" ")[2]
    dbName = dbName.split(";")[0]
    try:
        os.mkdir(dbName)
        print("Database", dbName, "created.")
    except:
        print("!Failed to create database %s because is already exists." %dbName)
    return

#DROP DATABASE
def DropDatabase(line):
    dbName = line.split(" ")[2]
    dbName = dbName.split(";")[0]
    try:
        os.rmdir(dbName)
        print("Database", dbName, "deleted.")
    except:
        print("!Failed to delete %s because it does not exist." %dbName)
    return

#USE
def Use(line):
    dbName = line.split(" ")[1]
    dbName = dbName.split(";")[0]
    try:
        os.chdir(dbName)
        print("Using database %s." %dbName)
    except:
        try:
            os.chdir("..")
            os.chdir(dbName)
            print("Using database ../%s." %dbName)
        except:
            print("!Cannot access database %s." %dbName)
    return

#CREATE TABLE
def CreateTable(line):
    tableName = line.split(" ")[2]
    paramaters = line.split(" ", 3)[3]
    paramaters = paramaters.strip().replace(";", "").replace(", ", "|")
    paramaters = paramaters[1:-1]
    if not (os.path.exists(tableName)):
        fileName = open(tableName, "w")
        fileName.write(paramaters)
        print("Table %s created." %tableName)
        fileName.close()
    else:
        print("!Failed to create table %s because it already exists." %tableName)
    return

#DROP TABLE
def DropTable(line):
    tableName = line.split(" ")[2]
    tableName = tableName.split(";")[0]
    if os.path.exists(tableName):
        os.remove(tableName)
        print("Table %s deleted." %tableName)
    else:
        print("!Failed to delete %s because it does not exist." %tableName)
    return

#ALTER TABLE
def AlterTable(line):
    tableName = line.split(" ")[2]
    parameters = line.split(" ", 4)[4]
    parameters = parameters.replace(";", "").replace(", ", "|")
    if(line.split(" ")[3] == "ADD"):
        if os.path.exists(tableName):
            fileName = open(tableName, "a+")
            fileName.write("|%s" %parameters)
            fileName.close()
            print("Table %s modified." %tableName)
        else:
            print("!Failed to modify table %s because it does not exist." %tableName)
    return

#SELECT * FROM (Query All)
def SelectFromTable(line):
    tableName = line.split(" ")[3]
    tableName = tableName.split(";")[0]
    if(line.upper().find(" * ") != -1):
        if os.path.exists(tableName):
            fileName = open(tableName, "r").read().splitlines()
            for each in fileName:
                print(each)
        else:
            print("!Failed to query table %s because it does not exist." %tableName)
    return

#SELECT (Query Specfic)
def SelectTable(line, cmd):
    cmd.PrevSQL = line.split(" ", 1)[1]
    cmd.PrevSQL = cmd.PrevSQL.strip().replace(" ", "").replace(",", "|")
    cmd.DataManipulation = cmdName.SELECT
    return

#INSERT INTO
def InsertIntoTable(line, cmd):
    tableName = line.split(" ")[2]
    paramaters = line.split(" ", 3)[3]
    paramaters = paramaters.split(";")[0].replace("\t", "").replace(" ", "").replace("'", "")
    if(paramaters.upper().split("(")[0] == "VALUES"):
        paramaters = paramaters.split("(")[1].replace(",", "|")
        paramaters = paramaters[:-1]
        if os.path.exists(tableName):
            fileName = open(tableName, "a+")
            fileName.write("\n%s" %paramaters)
            fileName.close()
            print("1 new record inserted.")
        else:
            print("!Failed to insert into table %s because it does not exist." %tableName)
    return

#UPDATE
def UpdateTable(line, cmd):
    tableName = line.split(" ", 1)[1]
    tableName = tableName.strip()
    if os.path.exists(tableName):
        cmd.TableName = tableName
        cmd.DataManipulation = cmdName.UPDATE
    else:
        print("!Failed to find table %s because it does not exist." %tableName)
    return

#SET
def SetTable(line, cmd):
    cmd.PrevSQL = line.split(" ", 1)[1]
    cmd.PrevSQL = cmd.PrevSQL.strip().replace("'", "")
    return

#FROM
def FromTable(line, cmd):
    tableName = line.split(" ", 1)[1]
    tableName = tableName.strip()
    if os.path.exists(tableName):
        cmd.TableName = tableName
    else:
        print("!Failed to find table %s because it does not exist." %tableName)
    return

#DELETE FROM
def DeleteFromTable(line, cmd):
    tableName = line.split(" ", 2)[2]
    tableName = tableName.strip()
    if os.path.exists(tableName):
        cmd.TableName = tableName
        cmd.DataManipulation = cmdName.DELETE
    else:
        print("!Failed to find table %s because it does not exist." %tableName)
    return

#WHERE
def WhereTable(line, cmd):
    if not os.path.exists(cmd.TableName):
        print("!Failed to find table %s because it does not exist." %cmd.TableName)
        return

    #Parse WHERE command
    whereStr = line.split(" ", 1)[1]
    whereStr = whereStr.strip().replace(";", "").replace("'","")

    #Open table for reading
    with open(cmd.TableName, "r") as file:
        data = file.readlines()

    #Set parameters in a list
    firstLine = data[0]
    parametersIndex = firstLine.count(" ")
    firstLine = firstLine.replace("|", " ")
    parameters = [""]
    parameters[0] = firstLine.split(" ")[0]
    for index in range(1, parametersIndex):
        parameters.append(firstLine.split(" ")[index * 2])
    for index in range(len(parameters)):
        if(str(whereStr.split(" ")[0]) == str(parameters[index])):
            parametersIndex = index
        if(str(cmd.PrevSQL.split(" ")[0]) == str(parameters[index])):
            setIndex = index
    count = 0

    #UPDATE only supports "="
    if(cmd.DataManipulation == cmdName.UPDATE):
        if(whereStr.split(" ")[1] == "="):
            for index, each in enumerate(data):
                if(index != 0):
                    if(each.split("|")[parametersIndex] == whereStr.split(" ")[2]):
                        data[index] = each.replace(each.split("|")[setIndex], cmd.PrevSQL.split(" ")[2]) + "\n"
                        count = count + 1
        #Print amount of records (>0) modified
        if(count == 1):
            print("1 record modified.")
        elif(count > 1):
            print("%i records modified." %count)
    #DELETE only suports "=", ">"
    elif(cmd.DataManipulation == cmdName.DELETE):
        if(whereStr.split(" ")[1] == "="):
            for index, each in enumerate(data):
                if(index != 0):
                    if(each.split("|")[parametersIndex] == whereStr.split(" ")[2]):
                        data[index] = ""
                        count = count + 1
        #assumes float values, may need error handling in the future
        elif(whereStr.split(" ")[1] == ">"):
            for index, each in enumerate(data):
                if(index != 0):
                    if(float(each.split("|")[parametersIndex]) > float(whereStr.split(" ")[2])):
                        data[index] = ""
                        count = count + 1
        #Print amount of records (>0) deleted
        if(count == 1):
            print("1 record deleted.")
        elif(count > 1):
            print("%i records deleted." %count)
    #SELECT only supports "!="
    elif(cmd.DataManipulation == cmdName.SELECT):
        queryStr = ""
        if(whereStr.split(" ")[1] == "!="):
            for index, each in enumerate(data):
                if(index != 0):
                    if(each.split("|")[parametersIndex] != whereStr.split(" ")[2]):
                        for x in range(0, cmd.PrevSQL.count("|")+1):
                            for y in range(0, len(parameters)):
                                if(cmd.PrevSQL.split("|")[x] == parameters[y]):
                                    queryStr += each.split("|")[y] + "|"
                                    if(y == len(parameters)):
                                        queryStr += "\n"
            print(data[0].strip())
            print(queryStr[:-1].strip())
    else:
        print("!Invalid use of the WHERE command.")

    with open(cmd.TableName, "w") as file:
        file.writelines(data)

    cmd.TableName = ""
    cmd.PrevSQL = ""
    cmd.DataManipulation = cmdName.NONE
    return